snuba/subscriptions/scheduler_consumer.py

```diff
@@ -0,0 +1,279 @@
+from arroyo.processing import StreamProcessor
+from arroyo.processing.strategies import ProcessingStrategy
+from arroyo.processing.strategies.abstract import ProcessingStrategyFactory
+from arroyo.synchronized import commit_codec
+from arroyo.types import Position
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.utils.metrics import MetricsBackend
+from snuba.utils.streams.configuration_builder import build_kafka_consumer_configuration
+from snuba.utils.types import Interval, InvalidRangeError
+
+logger = logging.getLogger(__name__)
+
+
+class MessageDetails(NamedTuple):
+    offset: int
+    orig_message_ts: datetime
+
+
+class Tick(NamedTuple):
+    partition: int
+    offsets: Interval[int]
+    timestamps: Interval[datetime]
+
+    def time_shift(self, delta: timedelta) -> Tick:
+        """
+        Returns a new ``Tick`` instance that has had the bounds of its time
+        interval shifted by the provided delta.
+        """
+        return Tick(
+            self.partition,
+            self.offsets,
+            Interval(self.timestamps.lower + delta, self.timestamps.upper + delta),
+        )
+
+
+class CommitLogTickConsumer(Consumer[Tick]):
+    """
+    A modified version of the `TickConsumer` that returns `ticks` constructed
+    from messages of a single commit log consumer rather than a synchronized consumer.
+
+    Like the `TickConsumer` the consumer returns a `Tick` message, which is derived
+    from the timestamps of the previous two messages associated with a partition.
+    The `Tick` now includes the partition number as the partition for the tick
+    is no longer the same as the partition of the message.
+
+    Since there is no synchronized consumer involved in this implementation,
+    this tick consumer decodes the commit log message in order to get the partition
+    number and timestamp corresponding to the original topic being followed by
+    the commit log.
+
+    The intervals being returned by this consumer should be equivalent to
+    those in the original TickConsumer/Synchronized consumer implementation.
+    """
+
+    def __init__(
+        self, consumer: Consumer[KafkaPayload], time_shift: Optional[timedelta] = None
+    ) -> None:
+        self.__consumer = consumer
+        self.__previous_messages: MutableMapping[Partition, MessageDetails] = {}
+        self.__time_shift = time_shift if time_shift is not None else timedelta()
+
+    def subscribe(
+        self,
+        topics: Sequence[Topic],
+        on_assign: Optional[Callable[[Mapping[Partition, int]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[Partition]], None]] = None,
+    ) -> None:
+        self.__consumer.subscribe(topics, on_assign=on_assign, on_revoke=on_revoke)
+
+    def unsubscribe(self) -> None:
+        self.__consumer.unsubscribe()
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[Message[Tick]]:
```

Comment:
Although they are similar in concept the implementation actually does have a fair number of differences. I don't think it's worth abstracting the logic since the other one will be removed with the current scheduler and so I prefer not to go to the effort of coming up with abstraction that ultimately does not provide that much value long term since we are only going to have this one after the roll out.

I will bring the tests and docs over now so we don't forget later.